src/librustc_typeck/astconv.rs

```diff
@@ -648,7 +656,71 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                                        params_present.insert(match param.kind {
+                                            GenericParamDefKind::Lifetime => ParamKindOrd::Lifetime,
+                                            GenericParamDefKind::Type { .. } => ParamKindOrd::Type,
+                                            GenericParamDefKind::Const => ParamKindOrd::Const,
+                                        });
+                                    });
+                                    let is_lifetime_present =
+                                        params_present.contains(&ParamKindOrd::Lifetime);
+                                    let is_types_present =
+                                        params_present.contains(&ParamKindOrd::Type);
+                                    let is_consts_present =
+                                        params_present.contains(&ParamKindOrd::Const);
+
+                                    Self::generic_arg_mismatch_err(
+                                        tcx.sess,
+                                        arg,
+                                        kind.descr(),
+                                        Some(&format!(
+                                            "reorder the arguments: {}{}{}: `<{}>`",
+                                            if is_lifetime_present { "lifetimes" } else { "" },
+                                            if is_types_present {
+                                                if is_lifetime_present {
+                                                    ", then types"
+                                                } else {
+                                                    "types"
+                                                }
+                                            } else {
+                                                ""
+                                            },
+                                            if is_consts_present
+                                                && tcx.sess.features_untracked().const_generics
+                                            {
+                                                if is_lifetime_present || is_types_present {
+                                                    ", then consts"
+                                                } else {
+                                                    "consts"
+                                                }
+                                            } else {
+                                                ""
+                                            },
+                                            ordered_params
+                                                .into_iter()
+                                                .filter_map(|param| {
+                                                    let param_name = param.name.to_string();
+                                                    if param_name == "Self" {
```

Comment:
We should be checking `kw::SelfUpper`, rather than a hard-coded string. Also, you can use `then_some` to avoid this if-statement.